tokio/src/sync/broadcast.rs

```diff
@@ -589,34 +598,31 @@ impl<T> Sender<T> {
             slot.val.with_mut(|ptr| unsafe { *ptr = value });
         }
 
-        // Release the slot lock before the tail lock
+        // Release the slot lock before notifying the receivers.
         drop(slot);
 
+        tail.notify_rx();
+
         // Release the mutex. This must happen after the slot lock is released,
         // otherwise the writer lock bit could be cleared while another thread
         // is in the critical section.
         drop(tail);
 
-        // Notify waiting receivers
-        self.notify_rx();
-
         Ok(rem)
     }
+}
 
-    fn notify_rx(&self) {
-        let mut curr = self.shared.wait_stack.swap(ptr::null_mut(), SeqCst) as *const WaitNode;
-
-        while !curr.is_null() {
-            let waiter = unsafe { Arc::from_raw(curr) };
-
-            // Update `curr` before toggling `queued` and waking
-            curr = waiter.next.with(|ptr| unsafe { *ptr });
+impl Tail {
+    fn notify_rx(&mut self) {
+        while let Some(mut waiter) = self.waiters.pop_back() {
+            // Safety: `waiters` lock is still held.
+            let waiter = unsafe { waiter.as_mut() };
 
-            // Unset queued
-            waiter.queued.store(false, SeqCst);
+            assert!(waiter.queued);
+            waiter.queued = false;
 
-            // Wake
-            waiter.waker.wake();
+            let waker = waiter.waker.take().unwrap();
+            waker.wake();
```

Comment:
Should we be notifying waiters outside the lock (possibly using the same strategy as the semaphore)? Or is that something to change later?

Reply:
It doesn't apply here as the receivers do not need to acquire the mutex so they will not be blocked.